File: apps/api-spring/src/main/java/com/jiandou/api/task/TaskExecutionCoordinator.java
package com.jiandou.api.task;

import com.jiandou.api.task.application.port.TaskQueuePort;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Component;

@Component
public class TaskExecutionCoordinator {

    private final TaskQueuePort taskQueuePort;
    private final TaskRepository taskRepository;

    public TaskExecutionCoordinator(TaskQueuePort taskQueuePort, TaskRepository taskRepository) {
        this.taskQueuePort = taskQueuePort;
        this.taskRepository = taskRepository;
    }

    public void enqueue(TaskRecord task, String stage, String event, String message) {
        String previousStatus = task.status;
        dequeue(task);
        task.status = "PENDING";
        task.errorMessage = "";
        task.finishedAt = null;
        task.isQueued = true;
        markActiveAttemptQueued(task);
        taskQueuePort.enqueue(task.id);
        touch(task);
        recordTrace(task, stage, event, message, "INFO", Map.of("queue_mode", true));
        recordStatusHistory(task, previousStatus, "PENDING", stage, event, message);
        recordQueueEvent(task, "enqueued", Map.of("stage", stage, "event", event, "message", message));
        taskRepository.save(task);
    }

    public void dequeue(TaskRecord task) {
        boolean wasQueued = task.isQueued || task.queuePosition != null;
        taskQueuePort.remove(task.id);
        task.isQueued = false;
        task.queuePosition = null;
        if (wasQueued) {
            recordQueueEvent(task, "removed", Map.of("queue_mode", true));
        }
        taskRepository.save(task);
    }

    public void recomputeQueuePositions(Collection<TaskRecord> tasks) {
        Map<String, Integer> positions = new LinkedHashMap<>();
        List<String> snapshot = queueSnapshot();
        for (int index = 0; index < snapshot.size(); index++) {
            positions.put(snapshot.get(index), index + 1);
        }
        for (TaskRecord item : tasks) {
            Integer position = positions.get(item.id);
            item.queuePosition = position;
            item.isQueued = position != null;
        }
    }

    public List<String> queueSnapshot() {
        return new ArrayList<>(taskQueuePort.snapshot());
    }

    public Map<String, Object> createAttempt(TaskRecord task, String triggerType, Map<String, Object> payload) {
        Map<String, Object> row = new LinkedHashMap<>();
        task.currentAttemptNo += 1;
        String attemptId = "att_" + UUID.randomUUID().toString().replace("-", "");
        Map<String, Object> safePayload = payload == null ? Map.of() : payload;
        task.activeAttemptId = attemptId;
        row.put("attemptId", attemptId);
        row.put("taskId", task.id);
        row.put("attemptNo", task.currentAttemptNo);
        row.put("triggerType", triggerType);
        row.put("status", "PENDING");
        row.put("queueName", "default");
        row.put("workerInstanceId", "");
        row.put("queueEnteredAt", null);
        row.put("queueLeftAt", null);
        row.put("claimedAt", null);
        row.put("startedAt", null);
        row.put("finishedAt", null);
        row.put("resumeFromStage", stringValue(safePayload.get("resumeFromStage")));
        row.put("resumeFromClipIndex", intValue(safePayload.get("resumeFromClipIndex"), 0));
        row.put("failureCode", "");
        row.put("failureMessage", "");
        row.put("payload", safePayload);
        task.attempts.add(0, row);
        touch(task);
        taskRepository.save(task);
        taskRepository.saveAttempt(task.id, row);
        return row;
    }

    public void markActiveAttemptQueued(TaskRecord task) {
        Map<String, Object> attempt = activeAttempt(task);
        if (attempt == null) {
            return;
        }
        String now = nowIso();
        attempt.put("status", "QUEUED");
        attempt.put("queueEnteredAt", now);
        attempt.put("queueLeftAt", null);
        attempt.put("claimedAt", null);
        attempt.put("startedAt", null);
        attempt.put("workerInstanceId", "");
        attempt.put("finishedAt", null);
        attempt.put("failureMessage", "");
        taskRepository.saveAttempt(task.id, attempt);
    }

    public void markActiveAttemptRunning(TaskRecord task, String workerInstanceId) {
        Map<String, Object> attempt = activeAttempt(task);
        if (attempt == null) {
            return;
        }
        String now = nowIso();
        attempt.put("status", "RUNNING");
        attempt.put("workerInstanceId", workerInstanceId == null ? "" : workerInstanceId);
        attempt.put("claimedAt", now);
        attempt.put("queueLeftAt", now);
        attempt.put("startedAt", now);
        taskRepository.saveAttempt(task.id, attempt);
        recordQueueEvent(task, "claimed", Map.of("workerInstanceId", workerInstanceId == null ? "" : workerInstanceId));
    }

    public void markActiveAttemptFinished(TaskRecord task, String status, String errorMessage) {
        Map<String, Object> attempt = activeAttempt(task);
        if (attempt == null) {
            return;
        }
        String now = nowIso();
        attempt.put("status", status);
        attempt.put("finishedAt", now);
        if (errorMessage != null && !errorMessage.isBlank()) {
            attempt.put("failureMessage", errorMessage);
        }
        taskRepository.saveAttempt(task.id, attempt);
        recordQueueEvent(task, status == null ? "finished" : status.toLowerCase(), Map.of(
            "status", status == null ? "" : status,
            "errorMessage", errorMessage == null ? "" : errorMessage
        ));
    }

    public void recordTrace(TaskRecord task, String stage, String event, String message, String level, Map<String, Object> payload) {
        Map<String, Object> row = new LinkedHashMap<>();
        row.put("traceId", "trace_" + java.util.UUID.randomUUID().toString().replace("-", ""));
        row.put("timestamp", nowIso());
        row.put("level", level);
        row.put("stage", stage);
        row.put("event", event);
        row.put("message", message);
        row.put("payload", payload);
        task.trace.add(row);
        taskRepository.saveTrace(task.id, row);
        touch(task);
        taskRepository.save(task);
    }

    public void recordStatusHistory(TaskRecord task, String previousStatus, String nextStatus, String stage, String event, String reason) {
        Map<String, Object> row = new LinkedHashMap<>();
        row.put("statusHistoryId", "sthis_" + java.util.UUID.randomUUID().toString().replace("-", ""));
        row.put("taskId", task.id);
        row.put("previousStatus", previousStatus);
        row.put("nextStatus", nextStatus);
        row.put("progress", task.progress);
        row.put("stage", stage);
        row.put("event", event);
        row.put("reason", reason);
        row.put("operator", "system");
        row.put("changedAt", nowIso());
        row.put("payload", Map.of());
        task.statusHistory.add(row);
        taskRepository.saveStatusHistory(task.id, row);
        taskRepository.save(task);
    }

    public void recordStageRun(TaskRecord task, Map<String, Object> stageRun) {
        task.stageRuns.add(stageRun);
        taskRepository.saveStageRun(task.id, stageRun);
        touch(task);
        taskRepository.save(task);
    }

    public void recordModelCall(TaskRecord task, Map<String, Object> modelCall) {
        task.modelCalls.add(modelCall);
        taskRepository.saveModelCall(task.id, modelCall);
        touch(task);
        taskRepository.save(task);
    }

    public void recordMaterial(TaskRecord task, Map<String, Object> material) {
        task.materials.add(material);
        taskRepository.saveMaterial(task.id, material);
        touch(task);
        taskRepository.save(task);
    }

    public void recordResult(TaskRecord task, Map<String, Object> result) {
        task.outputs.add(result);
        taskRepository.saveResult(task.id, result);
        touch(task);
        taskRepository.save(task);
    }

    public void recordQueueEvent(TaskRecord task, String eventType, Map<String, Object> payload) {
        Map<String, Object> row = new LinkedHashMap<>();
        row.put("taskQueueEventId", "queueevt_" + java.util.UUID.randomUUID().toString().replace("-", ""));
        row.put("taskId", task.id);
        row.put("attemptId", task.activeAttemptId == null ? "" : task.activeAttemptId);
        row.put("queueName", "default");
        row.put("eventType", eventType);
        row.put("workerInstanceId", activeAttemptWorkerId(task));
        row.put("queuePositionHint", task.queuePosition == null ? 0 : task.queuePosition);
        row.put("payload", payload == null ? Map.of() : payload);
        row.put("eventTime", nowIso());
        taskRepository.saveQueueEvent(task.id, row);
    }

    public void upsertWorkerInstance(String workerInstanceId, String workerType, String status, Map<String, Object> metadata) {
        String now = nowIso();
        Map<String, Object> row = new LinkedHashMap<>();
        row.put("workerInstanceId", workerInstanceId);
        row.put("workerType", workerType);
        row.put("queueName", "default");
        row.put("hostName", "");
        row.put("processId", ProcessHandle.current().pid());
        row.put("status", status);
        row.put("startedAt", now);
        row.put("lastHeartbeatAt", now);
        row.put("stoppedAt", "");
        row.put("metadata", metadata == null ? Map.of() : metadata);
        taskRepository.saveWorkerInstance(row);
    }

    public void touchWorkerInstance(String workerInstanceId, String workerType, String status, Map<String, Object> metadata) {
        Map<String, Object> existing = taskRepository.findWorkerInstance(workerInstanceId);
        String startedAt = existing == null ? nowIso() : String.valueOf(existing.getOrDefault("startedAt", nowIso()));
        Map<String, Object> row = new LinkedHashMap<>();
        row.put("workerInstanceId", workerInstanceId);
        row.put("workerType", workerType);
        row.put("queueName", String.valueOf(existing == null ? "default" : existing.getOrDefault("queueName", "default")));
        row.put("hostName", String.valueOf(existing == null ? "" : existing.getOrDefault("hostName", "")));
        row.put("processId", existing == null ? ProcessHandle.current().pid() : existing.getOrDefault("processId", ProcessHandle.current().pid()));
        row.put("status", status);
        row.put("startedAt", startedAt);
        row.put("lastHeartbeatAt", nowIso());
        row.put("stoppedAt", "STOPPED".equalsIgnoreCase(status) || "FAILED".equalsIgnoreCase(status) ? nowIso() : "");
        row.put("metadata", metadata == null ? (existing == null ? Map.of() : existing.getOrDefault("metadata", Map.of())) : metadata);
        taskRepository.saveWorkerInstance(row);
    }

    public int recoverStaleClaims(OffsetDateTime staleBefore, int limit) {
        int recovered = 0;
        for (String workerInstanceId : taskRepository.listStaleWorkerInstanceIds(staleBefore, Math.max(1, limit))) {
            markWorkerInstanceStale(workerInstanceId);
        }
        for (Map<String, Object> claim : taskRepository.listStaleRunningClaims(staleBefore, Math.max(1, limit))) {
            String taskId = String.valueOf(claim.getOrDefault("taskId", ""));
            if (taskId.isBlank()) {
                continue;
            }
            TaskRecord task = taskRepository.findById(taskId);
            if (task == null) {
                continue;
            }
            Map<String, Object> attempt = activeAttempt(task);
            if (attempt == null || !"RUNNING".equals(String.valueOf(attempt.getOrDefault("status", "")))) {
                continue;
            }
            String staleWorkerInstanceId = String.valueOf(claim.getOrDefault("workerInstanceId", ""));
            String previousStatus = task.status;
            task.status = "PENDING";
            task.progress = 0;
            task.errorMessage = "";
            task.finishedAt = null;
            task.isQueued = true;
            task.queuePosition = null;
            if (task.executionContext != null) {
                task.executionContext.put("recoveredFromWorkerInstanceId", staleWorkerInstanceId);
                task.executionContext.remove("workerInstanceId");
            }
            markActiveAttemptQueued(task);
            recordQueueEvent(task, "retry_enqueued", Map.of(
                "reason", "stale_claim_recovered",
                "staleWorkerInstanceId", staleWorkerInstanceId
            ));
            recordTrace(task, "dispatch", "task.recovered_from_stale_claim", "检测到失效 worker，任务已重新入队。", "WARN", Map.of(
                "staleWorkerInstanceId", staleWorkerInstanceId
            ));
            recordStatusHistory(task, previousStatus, "PENDING", "dispatch", "task.recovered_from_stale_claim", "检测到失效 worker，任务已重新入队。");
            touch(task);
            taskRepository.save(task);
            recovered += 1;
        }
        return recovered;
    }

    private void touch(TaskRecord task) {
        task.updatedAt = nowIso();
    }

    private String stringValue(Object value) {
        return value == null ? "" : String.valueOf(value).trim();
    }

    private int intValue(Object value, int fallback) {
        if (value instanceof Number number) {
            return number.intValue();
        }
        if (value != null) {
            try {
                return Integer.parseInt(String.valueOf(value).trim());
            } catch (NumberFormatException ignored) {
            }
        }
        return fallback;
    }

    private Map<String, Object> activeAttempt(TaskRecord task) {
        if (task.activeAttemptId == null || task.activeAttemptId.isBlank()) {
            return null;
        }
        for (Map<String, Object> row : task.attempts) {
            if (task.activeAttemptId.equals(row.get("attemptId"))) {
                return row;
            }
        }
        return null;
    }

    private String nowIso() {
        return OffsetDateTime.now(ZoneOffset.UTC).toString();
    }

    private String activeAttemptWorkerId(TaskRecord task) {
        Map<String, Object> attempt = activeAttempt(task);
        if (attempt == null) {
            return "";
        }
        Object value = attempt.get("workerInstanceId");
        return value == null ? "" : String.valueOf(value);
    }

    private void markWorkerInstanceStale(String workerInstanceId) {
        if (workerInstanceId == null || workerInstanceId.isBlank()) {
            return;
        }
        Map<String, Object> existing = taskRepository.findWorkerInstance(workerInstanceId);
        if (existing == null || existing.isEmpty()) {
            return;
        }
        if (!"RUNNING".equalsIgnoreCase(String.valueOf(existing.getOrDefault("status", "")))) {
            return;
        }
        Map<String, Object> row = new LinkedHashMap<>(existing);
        row.put("status", "STALE");
        row.put("stoppedAt", nowIso());
        taskRepository.saveWorkerInstance(row);
    }
}

File: backend/services/task_execution_coordinator.py
"""Task execution coordinator.

Translates the Java TaskExecutionCoordinator. Handles enqueue/dequeue,
attempt lifecycle, state transitions, trace/status-history recording, and
worker instance management.
"""

from __future__ import annotations

from collections.abc import Callable
from datetime import datetime
from typing import Any

from backend.domain.enums import (
    AttemptStatus,
    AttemptTriggerType,
    QueueEventType,
)
from backend.domain.task_record import TaskRecord
from backend.infrastructure.task_persistence_mutation import TaskPersistenceMutation
from backend.infrastructure.task_queue_port import InMemoryTaskQueue, TaskQueuePort
from backend.services.task_attempt_lifecycle import TaskAttemptLifecycle
from backend.services.task_attempt_mutation_service import TaskAttemptMutationService
from backend.services.task_execution_mutation_recorder import TaskExecutionMutationRecorder
from backend.services.task_execution_record_factory import TaskExecutionRecordFactory
from backend.services.task_queue_lifecycle import TaskQueueLifecycle
from backend.services.task_stale_claim_recovery import TaskStaleClaimRecovery
from backend.services.task_state_transition import TaskStateTransition, TaskStateTransitionBuilder
from backend.services.task_transition_service import TaskTransitionService
from backend.services.task_worker_registry import TaskWorkerRegistry


class TaskExecutionCoordinator:
    """Coordinates task execution lifecycle: queue, attempt, state transition."""

    def __init__(
        self,
        task_queue_port: TaskQueuePort | None = None,
    ) -> None:
        self._task_queue_port: TaskQueuePort = task_queue_port or InMemoryTaskQueue()
        self._attempt_lifecycle = TaskAttemptLifecycle()
        self._record_factory = TaskExecutionRecordFactory()
        self._mutation_recorder = TaskExecutionMutationRecorder(
            self._record_factory,
            self._attempt_lifecycle.active_worker_id,
        )
        self._attempt_mutations = TaskAttemptMutationService(
            self._attempt_lifecycle,
            self._mutation_recorder,
        )
        self._transition_service = TaskTransitionService(
            self._attempt_mutations,
            self._mutation_recorder,
        )
        self._queue_lifecycle = TaskQueueLifecycle(
            self._task_queue_port,
            self._attempt_lifecycle,
            self._record_factory,
        )
        self._worker_registry = TaskWorkerRegistry()
        self._stale_claim_recovery = TaskStaleClaimRecovery(self)

    @property
    def task_queue_port(self) -> TaskQueuePort:
        return self._task_queue_port

    # ------------------------------------------------------------------
    # Enqueue / Dequeue
    # ------------------------------------------------------------------

    def enqueue(
        self,
        task: TaskRecord,
        stage: str,
        event: str,
        message: str,
    ) -> dict[str, Any]:
        return self._queue_lifecycle.enqueue(task, stage, event, message)

    def dequeue(self, task: TaskRecord) -> dict[str, Any]:
        return self._queue_lifecycle.dequeue(task)

    def recompute_queue_positions(self, tasks: list[TaskRecord]) -> None:
        self._queue_lifecycle.recompute_positions(tasks)

    def queue_snapshot(self) -> list[str]:
        return self._queue_lifecycle.snapshot()

    # ------------------------------------------------------------------
    # Attempt lifecycle
    # ------------------------------------------------------------------

    def create_attempt(
        self,
        task: TaskRecord,
        trigger_type: str | AttemptTriggerType,
        payload: dict[str, Any] | None,
    ) -> dict[str, Any]:
        return self._attempt_mutations.create(task, trigger_type, payload)

    def mark_active_attempt_queued(self, task: TaskRecord) -> dict[str, Any] | None:
        return self._attempt_mutations.mark_queued(task)

    def mark_active_attempt_running(
        self,
        task: TaskRecord,
        worker_instance_id: str,
    ) -> dict[str, Any] | None:
        return self._attempt_mutations.mark_running(task, worker_instance_id)

    def mark_active_attempt_finished(
        self,
        task: TaskRecord,
        status: AttemptStatus | str,
        error_message: str | None = None,
    ) -> dict[str, Any] | None:
        return self._attempt_mutations.mark_finished(task, status, error_message)

    # ------------------------------------------------------------------
    # State transition
    # ------------------------------------------------------------------

    def transition_task(
        self,
        task: TaskRecord,
        transition: TaskStateTransition | TaskStateTransitionBuilder,
        task_mutator: Callable[[TaskRecord], None] | None = None,
    ) -> dict[str, Any]:
        return self._transition_service.transition(task, transition, task_mutator)

    def record_trace(
        self,
        task: TaskRecord,
        stage: str,
        event: str,
        message: str,
        level: str,
        payload: dict[str, Any] | None,
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_trace(task, stage, event, message, level, payload)

    def record_status_history(
        self,
        task: TaskRecord,
        previous_status: str,
        next_status: str,
        stage: str,
        event: str,
        reason: str,
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_status_history(
            task, previous_status, next_status, stage, event, reason
        )

    def record_stage_run(
        self,
        task: TaskRecord,
        stage_run: dict[str, Any],
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_stage_run(task, stage_run)

    def record_model_call(
        self,
        task: TaskRecord,
        model_call: dict[str, Any],
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_model_call(task, model_call)

    def record_material(
        self,
        task: TaskRecord,
        material: dict[str, Any],
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_material(task, material)

    def record_result(
        self,
        task: TaskRecord,
        result: dict[str, Any],
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_result(task, result)

    def record_queue_event(
        self,
        task: TaskRecord,
        event_type: str | QueueEventType,
        payload: dict[str, Any] | None,
    ) -> dict[str, Any]:
        return self._mutation_recorder.record_queue_event(task, event_type, payload)

    # ------------------------------------------------------------------
    # Worker instance management
    # ------------------------------------------------------------------

    def upsert_worker_instance(
        self,
        worker_instance_id: str,
        worker_type: str,
        status: str,
        metadata: dict[str, Any] | None,
    ) -> dict[str, Any]:
        row = self._worker_registry.upsert(worker_instance_id, worker_type, status, metadata)
        return {
            "mutation": TaskPersistenceMutation().add_worker_instance(row),
            "worker_instance": row,
        }

    def touch_worker_instance(
        self,
        worker_instance_id: str,
        worker_type: str,
        status: str,
        metadata: dict[str, Any] | None,
        existing_instance: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        row = self._worker_registry.touch(
            worker_instance_id,
            worker_type,
            status,
            metadata,
            existing_instance,
        )
        return {
            "mutation": TaskPersistenceMutation().add_worker_instance(row),
            "worker_instance": row,
        }

    async def recover_stale_claims(
        self,
        stale_before: datetime,
        limit: int,
        task_repository: Any,  # TaskRepository protocol
    ) -> int:
        return await self._stale_claim_recovery.recover(stale_before, limit, task_repository)

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _active_attempt(self, task: TaskRecord) -> dict[str, Any] | None:
        return self._attempt_lifecycle.active(task)

    def _active_attempt_worker_id(self, task: TaskRecord) -> str:
        return self._attempt_lifecycle.active_worker_id(task)

    def _mark_active_attempt_queued_in_memory(
        self,
        task: TaskRecord,
    ) -> dict[str, Any] | None:
        return self._attempt_lifecycle.mark_queued(task)

    def _touch(self, task: TaskRecord) -> None:
        self._mutation_recorder.touch(task)

    def _new_trace_row(
        self,
        stage: str,
        event: str,
        message: str,
        level: str,
        payload: dict[str, Any] | None,
    ) -> dict[str, Any]:
        return self._mutation_recorder.new_trace(stage, event, message, level, payload)

    def _new_status_history_row(
        self,
        task: TaskRecord,
        previous_status: str,
        next_status: str,
        stage: str,
        event: str,
        reason: str,
    ) -> dict[str, Any]:
        return self._mutation_recorder.new_status_history(
            task, previous_status, next_status, stage, event, reason
        )

    def _new_queue_event_row(
        self,
        task: TaskRecord,
        event_type: str | QueueEventType,
        payload: dict[str, Any] | None,
    ) -> dict[str, Any]:
        return self._mutation_recorder.new_queue_event(task, event_type, payload)

    def _apply_attempt_transition(
        self,
        task: TaskRecord,
        transition: TaskStateTransition,
    ) -> dict[str, Any] | None:
        return self._attempt_mutations.apply_transition(task, transition)
